# Business Understanding

## Overview



The Paris Olympics is a global sporting event that has garnered significant attention and engagement across various social media platforms. Analyzing public sentiment regarding the Olympics can provide valuable insights into how athletes, countries and the overall event are perceived. This analysis can benefit sports organizations, media outlets, sponsors offering feedback on public perception, performance and engagement levels thus helping to tailor content and marketing strategies. Sentiment analysis can also benefit city officials to improve planning and address concerns such as health and sanitation.
The goal of this project is to perform a comprehensive sentiment analysis of social media content related to this year's Paris Olympics to understand public sentiment, identify emerging trends and provide a comprehensive understanding of how different aspects of the Olympics resonate with audiences worldwide.

## Problem Statement



The Paris Olympics is a high-profile event that generates a substantial volume of unstructured social media data that reflects public sentiment. The challenge lies in effectively analyzing this vast and diverse stream of data while also tackling challenges such as language differences, sentiment variations and contextual meanings in order to provide accurate and actionable insights. 

## Proposed Solutions



1.	Use API access to collect data from major social media platforms and ensure compliance with platform policies and data protection regulations.
2.	Implement text normalization, tokenization and content filtering while utilizing language detection and translation tools for multilingual data handling. 
3.	Employ advanced natural language processing models like BERT or GPT for sentiment classification incorporating sarcasm detection and contextual analysis for improved accuracy. 
4.	Create an interactive dashboard using Tableau to display sentiment trends and insights with features for data filtering and exploring different aspects of the data.

## Objectives




### Main Objective
Develop a comprehensive social media sentiment analysis model that accurately captures and interprets public sentiment about the Paris Olympics from social media data.

### Specific Objectives
1.	To extract, preprocess and clean social media data from multiple platforms addressing quality issues and handling multilingual content related to the Paris Olympics. 
2.	To develop and train advanced natural language processing models to accurately classify sentiments incorporating techniques to handle sarcasm and contextual nuances. 
3.	To create interactive visualizations to display sentiment trends and key events providing actionable insights to stakeholders based on comprehensive analysis of public opinions.

## Success Metrics



- Accuracy – The proportion of correctly classified sentiments (positive, negative, neutral) out of all sentiments predicted by the model.
85% - 90%

- Precision -  The proportion of true positive sentiment predictions (correctly identified positive tweets) out of all predicted positives.
80% - 90% for both positive and negative sentiment classes.
75% - 85% for the neutral class.

- Recall - The proportion of true positive sentiment predictions out of all actual positives.
75% - 80% for all sentiment classes.

- F1 Score - The harmonic mean of Precision and Recall that provides a single metric that balances both precision and recall.
0.75 to 0.85

- Area Under the Curve - Receiver Operating Characteristic (AUC-ROC) - Measures how well a model distinguishes between classes. > 0.85

## Challenges


- Social media data is noisy and unstructured presenting challenges for accurate analysis. 
- Distinguishing between positive, negative and neutral sentiments can be difficult especially when dealing with multilingual content thus affecting sentiment analysis accuracy
- The volume of social media posts and comments can be overwhelming particularly during major events like the Olympics. Managing and processing large volumes of real-time data necessitates efficient data handling and processing techniques. 
- Interpreting context and sarcasm an extra layer of complexity as the sentiment expressed may not always align with the literal meaning of the words used. Social media content often includes informal language, slang and nuanced expressions that can skew sentiment analysis

## Conclusion



This sentiment analysis project aims to deliver a comprehensive understanding of public opinion about the Paris Olympics by leveraging social media data. By addressing the challenges of data quality, sentiment accuracy, multilingual content and implementing advanced NLP techniques, the project will provide actionable insights to the aforementioned stakeholders. Successful execution will enable better engagement strategies and enhance the overall experience of the Olympics for audiences worldwide.

## DATA UNDERSTANDING




### Data Sources
1.	APIs - Extract data from social media sites such as Twitter, Facebook and Instagram in the form of posts, tweets, comments and hashtags using their respective APIs. 
The focus will be on posts mentioning Paris Olympics, relevant hashtags and location-based data.

2.	Web Scraping - Extract additional data from comments and discussions from news sites and sports forums such as ESPN and Sports Center

### Datasets



1.	Social media data in the form of tweets, facebook and Instagram posts and comments mentioning the Paris Olympics.
2.	News articles, comments and replies discussing the various aspects of the Olympics.

### Relevance of The Data



>The data sources and datasets identified for this project are highly relevant to analyzing public sentiment surrounding the Paris Olympics. Social media platforms like Twitter, Facebook and Instagram capture immediate reactions, discussions and emotional responses from a global audience thus providing a rich source of unfiltered public sentiment. 
The inclusion of location-based data and relevant hashtags allows for more targeted analysis potentially revealing geographical trends and topic-specific sentiments. Complementing this with web scraping of news sites and sports forums like ESPN and Sports Center adds depth to the analysis by incorporating more structured discussions and content.
This combination of data sources offers a comprehensive view of public sentiment ranging from spontaneous reactions on social media to more considered opinions in news comment sections and sports forums.

#### Python Modules Importation

In [119]:
# import necessary modules
# manupulation
import pandas as pd
import os
import numpy as np

#### Combining CSV Files

In [130]:
# Folder in which the csv files are found
folder_path = 'X_data'

# List of all the csv files found in the folder
csv_files = [fil for fil in os.listdir(folder_path) if fil.endswith('.csv')]

# Create an empty list to store DataFrames
data_frames = []

# Loop through the list of CSV files and read each one into a DataFrame
for file in csv_files:
    file_path = os.path.join(folder_path, file)
    df = pd.read_csv(file_path)
    data_frames.append(df)

# Concatenate all DataFrames into one (stack vertically)
merged_df = pd.concat(data_frames, ignore_index=True, sort=False)

merged_df.info()

# Save the concatenated DataFrame to a new CSV file
new_df = '2024-olympics-sentiments.csv'
merged_df.to_csv(new_df, index=False)
# print(f'merged CSV file saved as {output_file}')



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12888 entries, 0 to 12887
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Category                  12888 non-null  object
 1   Keyword                   12888 non-null  object
 2   Web_Page_URL              12888 non-null  object
 3   Tweet_Website             12888 non-null  object
 4   Author_Name               12875 non-null  object
 5   Author_Web_Page_URL       12888 non-null  object
 6   Tweet_Timestamp           12258 non-null  object
 7   Tweet_Content             12884 non-null  object
 8   Tweet_Image_URL           6505 non-null   object
 9   Tweet_Video_URL           12888 non-null  object
 10  Tweet_AD                  12888 non-null  bool  
 11  Tweet_Number_of_Reviews   9161 non-null   object
 12  Tweet_Number_of_Retweets  10879 non-null  object
 13  Tweet_Number_of_Likes     12116 non-null  object
 14  Tweet_Number_of_Looks 

In [134]:
class DataUnderstanding():
    """Class that provides an understanding of a dataset"""
    
    def __init__(self, data=None):
        """Initialization"""
        self.df = data

    def load_data(self, path):
        """Load the data"""
        if self.df is None:
            self.df = pd.read_csv(path, encoding='latin-1')
        return self.df

    def understanding(self):
        """Provides insights into the dataset"""
        # Info
        print("INFO")
        print("-" * 4)
        self.df.info()

        # Shape
        print("\n\nSHAPE")
        print("-" * 5)
        print(f"Records in dataset: {self.df.shape[0]} with {self.df.shape[1]} columns.")

        # Columns
        print("\n\nCOLUMNS")
        print("-" * 6)
        print("Columns in the dataset are:")
        for idx in self.df.columns:
            print(f"- {idx}")

        # Unique Values
        print("\n\nUNIQUE VALUES")
        print("-" * 12)
        for col in self.df.columns:
            print(f"Column {col} has {self.df[col].nunique()} unique values")
            if self.df[col].nunique() < 12:
                print(f"Top unique values in {col} include:")
                for idx in self.df[col].value_counts().index:
                    print(f"- {idx}")
            print("")

        # Missing or Null Values
        print("\nMISSING VALUES")
        print("-" * 15)
        for col in self.df.columns:
            print(f"Column {col} has {self.df[col].isnull().sum()} missing values.")

        # Duplicate Values
        print("\n\nDUPLICATE VALUES")
        print("-" * 16)
        print(f"The dataset has {self.df.duplicated().sum()} duplicated records.")


In [140]:
# Create an instance of DataCleaning
data =  DataUnderstanding()

# Load the data (replace with your actual file path)
data_path = '2024-olympics-sentiments.csv'
df = data.load_data(data_path)
df

Category        Keyword  \
0      TopPost  #olympics2024   
1      TopPost  #olympics2024   
2      TopPost  #olympics2024   
3      TopPost  #olympics2024   
4      TopPost  #olympics2024   
...        ...            ...   
12883  TopPost     #paris2024   
12884  TopPost     #paris2024   
12885  TopPost     #paris2024   
12886  TopPost     #paris2024   
12887  TopPost     #paris2024   

                                            Web_Page_URL  \
0      https://twitter.com/search?q=#olympics2024&src...   
1      https://twitter.com/search?q=#olympics2024&src...   
2      https://twitter.com/search?q=#olympics2024&src...   
3      https://twitter.com/search?q=#olympics2024&src...   
4      https://twitter.com/search?q=#olympics2024&src...   
...                                                  ...   
12883  https://twitter.com/search?q=#paris2024&src=ty...   
12884  https://twitter.com/search?q=#paris2024&src=ty...   
12885  https://twitter.com/search?q=#paris2024&src=ty...   
12886  https://twitter.com/search?q=#paris2024&src=ty...   
12887  https://twitter.com/search?q=#paris2024&src=ty...   

                                           Tweet_Website  \
0      https://twitter.com/Mapasexy/status/1821618461...   
1      https://twitter.com/MuzalaSamukonga/status/182...   
2      https://twitter.com/gondo_nicole/status/182162...   
3      https://twitter.com/malinga_ninety9/status/182...   
4      https://twitter.com/LanceUlanoff/status/182018...   
...                                                  ...   
12883  https://twitter.com/Linus2k/status/88563519166...   
12884  https://twitter.com/ambafrancejp_jp/status/885...   
12885  https://twitter.com/rubencronicas/status/88602...   
12886  https://twitter.com/marcos_z/status/8848591646...   
12887  https://twitter.com/vpecresse/status/884862216...   

                              Author_Name  \
0                 Mapaseka MakhanyaReload   
1                        Muzala Samukonga   
2                          Nicole GReload   
3                          Lasith Malinga   
4                     Lance UlanoffReload   
...                                   ...   
12883                        Doctor Linus   
12884               フランス大使館   
12885                   RUBÉN HERNÁNDEZ   
12886      Marcos ZugastiReloadParis 2024   
12887  Valérie PécresseReloadParis 2024   

                       Author_Web_Page_URL Tweet_Timestamp  \
0             https://twitter.com/Mapasexy      2024-08-08   
1      https://twitter.com/MuzalaSamukonga      2024-08-08   
2         https://twitter.com/gondo_nicole      2024-08-08   
3      https://twitter.com/malinga_ninety9      2024-08-08   
4         https://twitter.com/LanceUlanoff      2024-08-04   
...                                    ...             ...   
12883          https://twitter.com/Linus2k      2017-07-13   
12884  https://twitter.com/ambafrancejp_jp      2017-07-14   
12885    https://twitter.com/rubencronicas      2017-07-15   
12886         https://twitter.com/marcos_z      2017-07-11   
12887        https://twitter.com/vpecresse      2017-07-11   

                                           Tweet_Content  \
0                                  He Said #Olympics2024   
1      ZAMBIA KU CHALO Thank you very much heavenly f...   
2      This was an African competition. Well done Afr...   
3      A huge applause to Arshad Nadeem and Neeraj Ch...   
4      Wow. Just wow. Noah Lyles wins by microns. #ol...   
...                                                  ...   
12883  Il mio augurio è che #Paris2024 sia un IMMENS...   
12884  今年もフランス大使å
¬é‚¸ã«ã¦ã€ãƒ•ã...   
12885  #Paris2024 @ClubFrance @FranciaenMexico @MX_Fr...   
12886                    Bancando los trapos. #Paris2024   
12887  Paris et @iledefrance accueilleront les Jeux O...   

                                         Tweet_Image_URL  \
0                                                    NaN   
1      https://pbs.twimg.com/media/GUev1

In [141]:

# Get an understanding of the dataset
data.understanding()

INFO
----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12888 entries, 0 to 12887
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Category                  12888 non-null  object
 1   Keyword                   12888 non-null  object
 2   Web_Page_URL              12888 non-null  object
 3   Tweet_Website             12888 non-null  object
 4   Author_Name               12875 non-null  object
 5   Author_Web_Page_URL       12888 non-null  object
 6   Tweet_Timestamp           12258 non-null  object
 7   Tweet_Content             12884 non-null  object
 8   Tweet_Image_URL           6505 non-null   object
 9   Tweet_Video_URL           12888 non-null  object
 10  Tweet_AD                  12888 non-null  bool  
 11  Tweet_Number_of_Reviews   9161 non-null   object
 12  Tweet_Number_of_Retweets  10879 non-null  object
 13  Tweet_Number_of_Likes     12116 non-null  object
 14  Tweet_Number

In [117]:
# Create an instance of DataUnderstanding
data = DataUnderstanding()

# Load the dataset
data_path = data.load_data(path="2024-olympics-sentiments.csv")

# Display the first five rows of the dataset
print(data_path.head())

# Get an understanding of the dataset
data.understanding()




  Category        Keyword                                       Web_Page_URL  \
0  TopPost  #olympics2024  https://twitter.com/search?q=#olympics2024&src...   
1  TopPost  #olympics2024  https://twitter.com/search?q=#olympics2024&src...   
2  TopPost  #olympics2024  https://twitter.com/search?q=#olympics2024&src...   
3  TopPost  #olympics2024  https://twitter.com/search?q=#olympics2024&src...   
4  TopPost  #olympics2024  https://twitter.com/search?q=#olympics2024&src...   

                                       Tweet_Website              Author_Name  \
0  https://twitter.com/Mapasexy/status/1821618461...  Mapaseka MakhanyaReload   
1  https://twitter.com/MuzalaSamukonga/status/182...         Muzala Samukonga   
2  https://twitter.com/gondo_nicole/status/182162...           Nicole GReload   
3  https://twitter.com/malinga_ninety9/status/182...           Lasith Malinga   
4  https://twitter.com/LanceUlanoff/status/182018...      Lance UlanoffReload   

                   Author_Web_Pa

## Data Cleaning

### Processes of Data Cleaning
- Dropping columns that are Irrelevant our analysis
- Removing duplicates
- Removing Null Values
- Stripping white spaces off the column names
- Convert the TweetTimestamp column into a datetime format 

In [144]:
import pandas as pd

class DataCleaning(DataUnderstanding):
    """This class is used for data cleaning"""
    
    def drop_columns(self, columns):
        """Drop specified columns"""
        print(f"Dropping Columns: {columns}")
        self.df.drop(columns=columns, inplace=True)

    def remove_duplicates(self):
        """Remove duplicates and print the number of duplicates removed"""
        num_duplicates_before = self.df.duplicated().sum()
        self.df.drop_duplicates(inplace=True)
        num_duplicates_after = self.df.duplicated().sum()
        print(f"Number of duplicates removed: {num_duplicates_before - num_duplicates_after}")

    def drop_nulls(self):
        """Drop rows with any null values and print the number of rows dropped"""
        num_rows_before = self.df.shape[0]
        self.df.dropna(inplace=True)
        num_rows_after = self.df.shape[0]
        print(f"Number of rows with null values dropped: {num_rows_before - num_rows_after}")
        
    def strip_column_names(self):
        """Strip whitespace from column names"""
        print("Stripping whitespace from column names")
        self.df.columns = self.df.columns.str.strip()
        
    def convert_to_datetime(self, column_name):
        """Convert a column to datetime format"""
        self.df[column_name] = pd.to_datetime(self.df[column_name])

    def filter_year(self, column_name, year):
        """Filter rows based on a specific year"""
        self.df = self.df[self.df[column_name].dt.year == year]

# Create an instance of DataCleaning
cleaner = DataCleaning()

# Load the data (replace with your actual file path)
data_path = '2024-olympics-sentiments.csv'
df = cleaner.load_data(data_path)

# Perform data cleaning
# Drop irrelevant columns
irrelevant_columns = ['Tweet_Image_URL', 'Web_Page_URL', 'Author_Web_Page_URL', 'Author_Name',
                      'Tweet_Video_URL', 'Tweet_AD', 'Tweet_Website', 'Tweet_Number_of_Reviews']
cleaner.drop_columns(irrelevant_columns)

# Remove duplicates and print the number removed
cleaner.remove_duplicates()

# Drop rows with null values and print the number removed
cleaner.drop_nulls()

# Strip column names of any leading/trailing whitespace
cleaner.strip_column_names()

# Convert 'Tweet_Timestamp' to datetime format
cleaner.convert_to_datetime('Tweet_Timestamp')

# Filter data for the year 2024
cleaner.filter_year('Tweet_Timestamp', 2024)

# Print the cleaned DataFrame
df = cleaner.df
print(df.head())


Dropping Columns: ['Tweet_Image_URL', 'Web_Page_URL', 'Author_Web_Page_URL', 'Author_Name', 'Tweet_Video_URL', 'Tweet_AD', 'Tweet_Website', 'Tweet_Number_of_Reviews']
Number of duplicates removed: 3206
Number of rows with null values dropped: 3665
Stripping whitespace from column names
  Category        Keyword Tweet_Timestamp  \
0  TopPost  #olympics2024      2024-08-08   
1  TopPost  #olympics2024      2024-08-08   
3  TopPost  #olympics2024      2024-08-08   
4  TopPost  #olympics2024      2024-08-04   
5  TopPost  #olympics2024      2024-08-08   

                                       Tweet_Content Tweet_Number_of_Retweets  \
0                              He Said #Olympics2024                      508   
1  ZAMBIA KU CHALO Thank you very much heavenly f...                      222   
3  A huge applause to Arshad Nadeem and Neeraj Ch...                       71   
4  Wow. Just wow. Noah Lyles wins by microns. #ol...                        4   
5  Welp looks like the goaltender for

**Validity**

Irrelevant Data; data that is not actually needed, and doesn’t fit under the context of the problem we’re trying to solve

***Why columns the columns have been dropped***

- `Web_Page_URL`: Redundant with the Tweet_Website column, because it contains the same information.
- `Author_Web_Page_URL`: It would be useful if we would have wanted to understand the who is posting but our focus is in the sentimental analysis
- `Tweet_Image_URL`:Images are not essential for this analysis considering the column itself has 50% null values
- `Tweet_Video_URL`:video content is not essential for this analysis.
- `Tweet_AD`: This indicates whether the tweet is an ad, butwe are not analyzing advertisements
- `Tweet_Number_of_Reviews`: Has a lot of missing values
 

**Completeness**

Data completeness is the extent to which all required and expected data elements are present within a dataset, ensuring that no essential information is missing.It is crucial for accurate insights, as missing information can lead to incomplete analyses and flawed decision-making. Neglecting data completeness risks distorted conclusions and undermines the reliability of business strategies or research outcomes.

In [150]:
print(df.isna().sum().sum())
 

0


- Number of rows with null values dropped: 3665

**Consistency**

In [151]:
print(df.duplicated().sum())

0


Number of duplicates removed: 3206